lib/path/default.nix

```
@@ -137,6 +141,84 @@ in /* No rec! Add dependencies on this file at the top. */ {
     subpathInvalidReason value == null;
 
 
+  /* Join subpath strings together using `/`, returning a normalised subpath string.
+
+    Like `concatStringsSep "/"` but safer, specifically:
+
+    - All elements must be valid subpath strings, see `lib.path.subpath.isValid`
+
+    - The result gets normalised, see `lib.path.subpath.normalise`
+
+    - The edge case of an empty list gets properly handled by returning the neutral subpath `"./."`
+
+    Laws:
+
+    - (Associativity):
+
+          subpath.join [ x (subpath.join [ y z ]) ] == subpath.join [ (subpath.join [ x y ]) z ]
+
+    - (Identity) The neutral "./." value is the neutral identity for normalised paths:
+
+          subpath.join [ ] == "./."
+          subpath.join [ (subpath.normalise p) "./." ] == subpath.normalise p
+          subpath.join [ "./." (subpath.normalise p) ] == subpath.normalise p
+
+    - (Normalised) The result is normalised according to `subpath.normalise`:
+
+          subpath.join ps == subpath.normalise (subpath.join ps)
+
+    - For non-empty lists, the implementation is equivalent to normalising the result of `concatStringsSep "/"`.
+      Note that the above laws can be derived from this one.
+
+          ps != [] -> subpath.join ps == subpath.normalise (concatStringsSep "/" ps)
+
+    Type:
+      subpath.join :: [ String ] -> String
+
+    Example:
+      subpath.join [ "foo" "bar/baz" ]
+      => "./foo/bar/baz"
+
+      # normalise the result
+      subpath.join [ "./foo" "." "bar//./baz/" ]
+      => "./foo/bar/baz"
+
+      # passing an empty list results in the current directory
+      subpath.join [ ]
+      => "./."
+
+      # elements must be valid subpath strings
+      subpath.join [ /foo ]
+      => <error>
+      subpath.join [ "" ]
+      => <error>
+      subpath.join [ "/foo" ]
+      => <error>
+      subpath.join [ "../foo" ]
+      => <error>
+  */
+  subpath.join = paths:
+    assert assertMsg
+      # Fast in case all paths are valid
+      (all (path: subpathInvalidReason path == null) paths)
+      # Otherwise we take our time to gather more info for a better error message
+      (let
+        # Strictly go through each path, throwing on the first invalid one
+        # Tracks the list index in the fold accumulator
+        trigger = foldl' (i: path:
+          if subpathInvalidReason path != null
+          then throw "lib.path.subpath.join: Element at index ${toString i} is not a valid subpath string: ${subpathInvalidReason path}"
```

```suggestion
          then throw ''
            lib.path.subpath.join:
              Element at index ${toString i} is not a valid subpath string:
                ${subpathInvalidReason path}
          ''
```

Since those strings are fairly long, I think we should break them down to roughly fit into a standard terminal.
No strong opinion on indentation, but it *may* help with debugging by indicating level of detail.

---

See https://github.com/NixOS/nixpkgs/pull/208887#discussion_r1069737602